src/time-to-interactive/TTIBoomerang.ts

```diff
@@ -0,0 +1,331 @@
+        if (this.fcpSupported) {
+            onFCP((metric) => this.handleWebVitals(metric));
+        }
+        if (this.lcpSupported) {
+            onLCP((metric) => this.handleWebVitals(metric));
+        }
+
+        // eslint-disable-next-line @typescript-eslint/no-empty-function
+        if (window.requestAnimationFrame(() => {})) {
+            // Init FPS listener if supported
+            this.fpsSupported = true;
+            this.framesPerSecondListener();
+        }
+    }
+
+    private computeTimeWindow(currTime?: number) {
+        if (typeof currTime === 'undefined') {
+            return Math.floor(performance.now() / this.COLLECTION_PERIOD);
+        } else {
+            return Math.floor(currTime / this.COLLECTION_PERIOD);
+        }
+    }
+
+    private addToTracker(type: string, bucket: number, value: any) {
+        if (!this.ttiTracker[type]) {
+            this.ttiTracker[type] = [];
+        }
+
+        if (!this.ttiTracker[type][bucket]) {
+            this.ttiTracker[type][bucket] = 0;
+        }
+
+        this.ttiTracker[type][bucket] += value;
+    }
+
+    eventListener = () => {
+        const eventObserver: PerformanceObserver = new PerformanceObserver(
+            (list) => {
+                list.getEntries()
+                    .filter(
+                        (e) =>
+                            e.entryType === LONG_TASK ||
+                            e.entryType === NAVIGATION
+                    )
+                    .forEach((event) => {
+                        if (event.entryType === LONG_TASK) {
+                            // Add to the time buckets where the long task spreads over
+                            if (event.startTime && event.duration) {
+                                const endTime =
+                                    event.startTime + event.duration;
+                                this.addToTracker(
+                                    LONG_TASK,
+                                    this.computeTimeWindow(event.startTime),
+                                    1
+                                );
+                                this.addToTracker(
+                                    LONG_TASK,
+                                    this.computeTimeWindow(endTime),
+                                    1
+                                );
+                            }
+                        }
+                        if (event.entryType === NAVIGATION) {
+                            if (event.toJSON().domContentLoadedEventEnd) {
+                                this.domContentLoadedEventEnd =
+                                    event.toJSON().domContentLoadedEventEnd;
+                            }
+                        }
+                    });
+            }
+        );
+        eventObserver.observe({
+            entryTypes: [LONG_TASK, NAVIGATION]
+        });
+    };
+
+    /*
+    Track frames per second 
+    */
+    framesPerSecondListener = () => {
```

Comment:
question: What does this do, and is it necessary? It seems like it might impact application performance.

Reply:
This listener tracks FPS per 100 ms time window by incrementing a time bucket on every animation frame. We need this to calculate the quiet window, which is defined by no long tasks and FPS > 20.

It looks like requestAnimationFrame is a standard way of calculating FPS at runtime, not sure what alternate methods are more performant.

Reply:
To improve performance, we could do a sample of FPS instead of calculating an exact value. But the current collection period is already fairly low at 100ms, so I would not recommend that

Reply:
Another performance improvement is that we could wait until after visuallyReadyTimestamp is calculated before tracking FPS

Reply:
@williazz summed it up accurately. It's one component that is part of the criteria to TTI. Right now this is the most performant way to gather this data. There is some [experimental](https://udn.realityripple.com/docs/Web/API/PerformanceFrameTiming) proposals around adding this as something that can be tracked by the PerformanceObserver, but there is no ETA on the feature. 

> Another performance improvement is that we could wait until after visuallyReadyTimestamp is calculated before tracking FPS

Experimented with this a bit before. Often times, if TTI is soon after VR (visually ready), the tracker does not have enough FPS datapoints leading to the TTI being delayed more than what it should be. I believe the tracker is in practice not instantaneous in initialization, and this leads to this problem. 

In most cases, it should have little impact on the [application](https://akamai.github.io/boomerang/oss/BOOMR.plugins.Continuity.html#toc5__anchor), but I can see cases where it might. 


One option is we can remove FPS from TTI computation entirely for now. In essence a quiet window with no long tasks should have an ideal frame rate so perhaps we are more or less covering this aspect of the measurement through the long task measurements already.

Reply:
Got it, thanks! There's a lot of good information here that should be persisted. Please add this info to the source file, including (1) a natural language description of the behavior, and (2) expected performance impact.

> One option is we can remove FPS from TTI computation entirely for now

We could make it optional and off by default. If (A) this is not critical to measuring TTI, and (B) we are unsure about the performance, this might be the right decision.

Reply:
added description and link to this discussion at top of `TimeToInteractive.ts`